crypto/dilithium.go

```diff
@@ -0,0 +1,175 @@
+package crypto
+
+import (
+	"fmt"
+
+	"github.com/cloudflare/circl/sign/dilithium"
+	"github.com/pkg/errors"
+)
+
+type (
+	DilithiumMode string
+)
+
+func (m DilithiumMode) String() string {
+	return string(m)
+}
+
+const (
+	Dilithium2 DilithiumMode = "Dilithium2"
+	Dilithium3 DilithiumMode = "Dilithium3"
+	Dilithium5 DilithiumMode = "Dilithium5"
+)
+
+func GenerateDilithiumKeyPair(m DilithiumMode) (dilithium.PublicKey, dilithium.PrivateKey, error) {
+	mode := dilithium.ModeByName(m.String())
+	if mode == nil {
+		return nil, nil, fmt.Errorf("unsupported dilithium mode: %s", m)
+	}
+	pk, sk, err := mode.GenerateKey(nil)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "could not generate key for dilithium")
+	}
+	return pk, sk, nil
+}
+
+type DilithiumSigner struct {
+	kid  string
+	mode dilithium.Mode
+	dilithium.PublicKey
+	dilithium.PrivateKey
+}
+
+func NewDilithiumSigner(kid string, m DilithiumMode, privKey dilithium.PrivateKey) (*DilithiumSigner, error) {
+	mode := dilithium.ModeByName(m.String())
+	if mode == nil {
+		return nil, fmt.Errorf("unsupported dilithium mode: %s", m)
+	}
+	// verify the mode matches the pk, this will panic if they don't match
+	publicKey := privKey.Public().(dilithium.PublicKey)
+	pubKey := mode.PublicKeyFromBytes(publicKey.Bytes())
+	return &DilithiumSigner{
+		kid:        kid,
+		mode:       mode,
+		PublicKey:  pubKey,
+		PrivateKey: privKey,
+	}, nil
+}
+
+func (s *DilithiumSigner) GetKeyID() string {
+	return s.kid
+}
+
+func (s *DilithiumSigner) Sign(message []byte) []byte {
+	return s.mode.Sign(s.PrivateKey, message)
+}
+
+type DilithiumVerifier struct {
+	KID  string
+	mode dilithium.Mode
+	dilithium.PublicKey
+}
+
+func NewDilithiumVerifier(kid string, m DilithiumMode, pubKey dilithium.PublicKey) (*DilithiumVerifier, error) {
+	mode := dilithium.ModeByName(m.String())
+	if mode == nil {
+		return nil, fmt.Errorf("unsupported dilithium mode: %s", m)
+	}
+	// verify the mode matches the pk, this will panic if they don't match
+	_ = mode.PublicKeyFromBytes(pubKey.Bytes())
+	return &DilithiumVerifier{
+		KID:       kid,
+		mode:      mode,
+		PublicKey: pubKey,
+	}, nil
+}
+
+func (s *DilithiumVerifier) Verify(message, signature []byte) bool {
+	return s.mode.Verify(s.PublicKey, message, signature)
+}
+
+func GetModeFromDilithiumPrivateKey(privKey dilithium.PrivateKey) (DilithiumMode, error) {
+	if err := getMode2PrivateKey(privKey); err == nil {
+		return Dilithium2, nil
+	}
+	if err := getMode3PrivateKey(privKey); err == nil {
+		return Dilithium3, nil
+	}
+	if err := getMode5PrivateKey(privKey); err == nil {
+		return Dilithium5, nil
+	}
+	return "", errors.New("unsupported dilithium mode")
+}
+
+func getMode2PrivateKey(privKey dilithium.PrivateKey) (err error) {
+	defer func() {
+		if r := recover(); r != nil {
+			err = errors.New("not mode 2")
+		}
+	}()
```

Comment:
Here and elsewhere - I would recommend doing the check that's causing the panic. It would avoid recovering from a potentially really bad situation (like OOM for example).

Reply:
can you expand on how I'd do this, since the library throws a panic and does not return an error how would we avoid it?

Reply:
do you mean move

```go
	if len(data) != mode2.PrivateKeySize {
		panic("packed public key must be of mode2.PrivateKeySize bytes")
	}
```

from the library in here and then forget about panics?

Reply:
I will do this but generally I don't like intercepting what a method from a lib will do since it could change out from underneath us